Factor refcount debugging code into a separate base class

geoffreygaren opened the pull request:
#### a978fae61986c0d3056e523f6552f28fc7b19b9f
<pre>
Factor refcount debugging code into a separate base class
<a href="https://bugs.webkit.org/show_bug.cgi?id=303029">https://bugs.webkit.org/show_bug.cgi?id=303029</a>
<a href="https://rdar.apple.com/165313018">rdar://165313018</a>

Reviewed by Chris Dumez.

Created a new base class, RefCountDebugger, to hold debug-only code. Moved
debug-only refcounting code into it.

This makes the refcounting logic more visible, removes a bunch of duplicate
code, and makes it easier to add a new refcounting class -- which I plan to do
soon.

Anything you&apos;re not testing is broken, so adding adoption assertions to
ThreadSafeRefCounted&lt;T&gt; uncovered some non-conforming code. I added
relaxAdoptionRequirement() to the ThreadSafeRefCounted&lt;T&gt; constructor to
maintain the status quo for now -- but we should really go back and fix this.

I added a create() function to RemoteInspectorXPCConnection before I gave up on
fixing adoption assertions, so I&apos;ve kept that in the patch.

Canonical link: <a href="https://commits.webkit.org/303536@main">https://commits.webkit.org/303536@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c88f857cc4b9550df9e2254202f3e0fa7d57ecd6

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/132691 "6 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/5186 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/43763 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/140227 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/84718 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/aaa94873-26de-4f04-a982-eb8fc3ff5cf0) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/134561 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/155/builds/5535 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/4945 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/101456 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/68747 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/0b0eb2e2-832b-4ff4-ad97-53f139674b03) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/135637 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/3832 "Unexpected infrastructure issue, retrying build") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/118870 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/82249 "Passed tests") | | ⏳ 🛠 vision-apple 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/3717 "Passed tests") | [❌ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/1450 "Found 1 new API test failure: TestWebKitAPI.ObscuredContentInsets.TopOverhangColorExtensionLayer (failure)") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/83453 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/124761 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/112672 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/36985 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/142874 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/131199 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/4856 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/37573 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/109828 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/4938 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/4202 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/110005 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/3705 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/115141 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/58336 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20586 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/4910 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/33489 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/164166 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/4748 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/68361 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/42635 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/5001 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/4867 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/JavaScriptCore/bytecode/Watchpoint.cpp:
```diff
@@ -98,6 +98,10 @@ WatchpointSet::WatchpointSet(WatchpointState state)
 WatchpointSet::~WatchpointSet()
 {
+    // FIXME: This is here to silence a RefcountDebugger ASSERT. But the ASSERT
+    // is correct and our code is incorrect!
```
We should probably file a bug/radar about that?

cdumez commented on Source/WTF/wtf/RefCountDebugger.h:
```diff
@@ -0,0 +1,188 @@
+// This base class holds debugging code, to share among refcounting subclasses.
+class RefCountDebugger {
+public:
+    enum class RefCountIsThreadSafe { No, Yes };
```
We usually use `enum class RefCountIsThreadSafe : bool { No, Yes };`

cdumez commented on Source/WTF/wtf/RefCountDebugger.h:
```diff
@@ -0,0 +1,188 @@
+    enum class RefCountIsThreadSafe { No, Yes };
+
+    WTF_EXPORT_PRIVATE static void logRefDuringDestruction(const void*);
+    WTF_EXPORT_PRIVATE static void printRefDuringDestructionLogAndCrash [[noreturn]] (const void*);
```
Don't we usually use `NO_RETURN_DUE_TO_CRASH` for such things?

cdumez commented on Source/WTF/wtf/RefCounted.h:
```diff
@@ -20,192 +20,63 @@
-
-protected:
-    RefCounted() { }
-    ~RefCounted()
```
Why isn't the destructor protected anymore? This class is not virtual so we should never be using polymorphism and calling this destructor directly. The previous code seemed safer.

cdumez commented on Source/WTF/wtf/ThreadSafeRefCounted.h:
```diff
@@ -34,103 +34,57 @@
-        return refCount() == 1;
+        // FIXME: Lots of subclasses violate our adoption requirements. Migrate
+        // this call into only those subclasses that need it.
+        relaxAdoptionRequirement();
```
Honestly, I don't feel like the adoption requirements bring much value nowadays and we have to keep relaxing them nowadays when doing Safer CPP adoption. Maybe we should just get rid of adoption requirements instead of trying to fix them.

geoffreygaren replied:
Haha, I thought the same thing until this found three cases where clients were just not doing refcounting right at all. So now I'm not so sure.

cdumez commented on Source/WTF/wtf/RefCounted.h:
```diff
@@ -20,192 +20,63 @@
+    WTF_MAKE_NONCOPYABLE(RefCounted);
+    WTF_DEPRECATED_MAKE_FAST_ALLOCATED(RefCounted);
 public:
+    RefCounted() = default;
```
Does this really need to be public? I doubt it. I would have kept it protected.

cdumez commented on Source/WTF/wtf/ThreadSafeRefCounted.h:
```diff
@@ -34,103 +34,57 @@
-
 template<class T, DestructionThread destructionThread = DestructionThread::Any> class ThreadSafeRefCounted : public ThreadSafeRefCountedBase {
 public:
+    ThreadSafeRefCounted() = default;
```
Can this be protected?

cdumez commented on Source/WTF/wtf/ThreadSafeRefCounted.h:
```diff
@@ -146,13 +100,9 @@ template<class T, DestructionThread destructionThread = DestructionThread::Any>
     }
-
-protected:
-    ThreadSafeRefCounted() = default;
```
Can we mark the destructor as protected?

cdumez commented on Source/WebKit/UIProcess/API/glib/WebKitWebContext.cpp:
```diff
@@ -439,23 +439,23 @@ static void webkitWebContextConstructed(GObject* object)
-    API::ProcessPoolConfiguration configuration;
-    configuration.setInjectedBundlePath(FileSystem::stringFromFileSystemRepresentation(bundleFilename.get()));
-    configuration.setUsesWebProcessCache(true);
+    auto configuration = API::ProcessPoolConfiguration::create();
```
Nowadays we like to be explicit about smart pointer and usually use `Ref` instead of `auto`.

cdumez reviewed (approved):
LGTM with suggestions.

geoffreygaren committed:
Factor refcount debugging code into a separate base class
https://bugs.webkit.org/show_bug.cgi?id=303029
rdar://165313018

Reviewed by Chris Dumez.

Created a new base class, RefCountDebugger, to hold debug-only code. Moved
debug-only refcounting code into it.

This makes the refcounting logic more visible, removes a bunch of duplicate
code, and makes it easier to add a new refcounting class -- which I plan to do
soon.

Anything you're not testing is broken, so adding adoption assertions to
ThreadSafeRefCounted<T> uncovered some non-conforming code. I added
relaxAdoptionRequirement() to the ThreadSafeRefCounted<T> constructor to
maintain the status quo for now -- but we should really go back and fix this.

I added a create() function to RemoteInspectorXPCConnection before I gave up on
fixing adoption assertions, so I've kept that in the patch.

Canonical link: https://commits.webkit.org/303536@main
